### PR TITLE
Document exceptions in Utilities

### PR DESCRIPTION
**Pull request by djthegr8**
## What does this PR do?
Adds documentation to exceptions in the Utilities folder.

Fixes a part of #5434

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you verify new and existing tests pass locally with your changes?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Yes 😂

### CODE REVIEW

File: pytorch_lightning/utilities/device_parser.py [hunk 1]
```diff
@@ -147,6 +157,10 @@ def _sanitize_gpu_ids(gpus: List[int]) -> List[int]:
 
     Returns:
         unmodified gpus variable
+
+    Raises:
+        MisconfigurationException:
+            If machine has lesser available GPUs than requested.
```
Comment (ananthsub) [inline pytorch_lightning/utilities/device_parser.py:163]:
```suggestion
            If machine has fewer available GPUs than requested.
```

File: pytorch_lightning/utilities/device_parser.py [hunk 2]
```diff
@@ -29,6 +29,12 @@ def determine_root_gpu_device(gpus: List[int]) -> Optional[int]:
 
     Returns:
         designated root GPU device id
+
+    Raises:
+        TypeError:
+            If GPUs are not in a list
+        AssertionError:
+            If GPU List is empty
```
Comment (ananthsub) [inline pytorch_lightning/utilities/device_parser.py:37]:
```suggestion
            If GPU list is empty
```

File: pytorch_lightning/utilities/device_parser.py [hunk 3]
```diff
@@ -29,6 +29,12 @@ def determine_root_gpu_device(gpus: List[int]) -> Optional[int]:
 
     Returns:
         designated root GPU device id
+
+    Raises:
+        TypeError:
+            If GPUs are not in a list
```
Comment (ananthsub) [inline pytorch_lightning/utilities/device_parser.py:35]:
```suggestion
            If ``gpus`` is not a List
```

though i don't get why we're raising an exception if the typehint already lists that this should be List[int]

Comment (djthegr8) [reply pytorch_lightning/utilities/device_parser.py:35]: type hint doesn't enforce so...

**Review by ananthsub — APPROVED** [review]
LGTM

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/data.py [hunk 4]
```diff
@@ -24,8 +24,14 @@ def has_iterable_dataset(dataloader: DataLoader):
 
 
 def has_len(dataloader: DataLoader) -> bool:
-    """ Checks if a given Dataloader has __len__ method implemented i.e. if
-    it is a finite dataloader or infinite dataloader. """
+    """
+    Checks if a given Dataloader has __len__ method implemented i.e. if
```
Comment (awaelchli) [inline pytorch_lightning/utilities/data.py:28]:
```suggestion
    Checks if a given Dataloader has ``__len__`` method implemented i.e. if
```

File: pytorch_lightning/utilities/device_parser.py [hunk 5]
```diff
@@ -29,6 +29,12 @@ def determine_root_gpu_device(gpus: List[int]) -> Optional[int]:
 
     Returns:
         designated root GPU device id
+
+    Raises:
+        TypeError:
+            If ``gpus`` is not a List
```
Comment (awaelchli) [inline pytorch_lightning/utilities/device_parser.py:35]:
```suggestion
            If ``gpus`` is not a list
```

**Review by awaelchli — APPROVED** [review]
great!

**Pull request merged**